### ISSUE TITLE
How might I implement a "last_up_to_depth()" behavior?

### ISSUE THREAD

**Issue by davidkrider** [stefankroes/ancestry#188]
I am trying to achieve a classic bill of materials behavior with revision levels. I need to get all the latest versions of each object up to a particular level. For instance, I may have 98 objects on the bill, many of which get revised many times. Say, at revision 23, I still may have 54 objects at revision 1, 19 objects at rev 2, on up to, say, 3 objects that got revised from version 22 to 23. So if I need to fetch the BOM at revision 17, I need all the latest children of each object up to rev 17. With the existing "to_depth()" method, it returns all the revisions of each object. I just want the latest copy of each. Something named like "last_up_to_depth()" makes sense to my mind. Is there already a way I could implement this with stock ActiveRecord behavior? If not, how hard would it be to patch ancestry with this method myself? My fear is that I may as well just write hard code myself, and not update the gem, because the performance of this behavior would stink, and wouldn't be worth putting in the gem. I can already do a "find_by_sql" that's very fast (and messy), so I'm just trying to estimate the proper way to go about this.

**Comment by kbrock** [comment]
Hello, this issue is quite old.

I realize that one reason it did not get proper attention is because it is a little hard to follow.

Do you still have this issue?
Did you end up going with a gem that is intended on providing versioning instead of parent/child?
Could you show a model or something so others can understand how the BOM (bill of materials) uses ancestry and how that ties with versioning?

I'm assuming you already have a solution to your problem, but if you are not happy with that solution, please share and we can hopefully get you super happy with your solution.

**Comment by davidkrider** [comment]
Yeah, it definitely is hard to follow. I ended up leveling up my SQL abilities, and doing it in the query in a "latest by date" sort of way. (The whole thing is here: https://dba.stackexchange.com/questions/94046/subselect-of-maxversion-takes-many-minutes-though-there-are-only-20k-records) This broke the ability to make a nice, clean AR call, since now it's about 50 lines of SQL, with the substitution of the revision level, but it's FAST, and AR still understands the objects in the returned results, so it was an acceptable tradeoff.

**Comment by kbrock** [comment]
@davidkrider

Thank you for your response and glad you found a great solution.

Windowing functions are pretty sweet, but they do tend to require a certain level of expertise to write and maintain them. In general, Rails developers tend to not be at that level.
Rails does now support Common Table Expressions, so hopefully you will be able to leverage that and trim down your sql.


Since it sounds like you solved this issue, I'm going to close this one out.
Please open other issues if there are requests that will help you with this and other features.

**Comment by davidkrider** [comment]
Nice. I've been trying to keep an eye on Rails 7, but I hadn't seen that feature. Thanks.
